opm/autodiff/BlackoilAquiferModel.hpp

```diff
@@ -0,0 +1,181 @@
+/*
```

Comment:
does this really need to be a separate class? it would probably better to integrate this into BlackOilModelEbos, so that the standard simulator supports this out of the box. Then again, maybe not...

Reply:
Also, it seems like this is based on the `flow_legacy` code. are there any reasons for that?

Reply:
@andlaus I am taking the structure from the BlackoilWellModel. I am not aware which is flow_legacy style or whatnot. If we can have a discussion on this, then I would gladly change to the updated functions etc...

Reply:
ah, right. I somehow messed this up. You will probably still need a separate `flow` binary if you add a new model, so it would be preferable to add aquifier support to the standard model, i.e. `BlackoilModelEbos.hpp`

Reply:
@andlaus  The aquifer model is indeed called within BlackoilModelEbos.hpp. This also means that there is an extra argument for the BlackoilModelEbos constructor and changes to the SimulatorFullyImplicitBlackoilEbos too. I anticipate that we might need such organization if we were to extend to multiple types of aquifers in the future.

Reply:
I don't want to call the shots here, but I generally prefer "one size fits all" solutions because they avoid tend to mitigate combinatorial explosions of possibilities while usually only making the base code slightly more complicated.

Reply:
@andlaus Not a problem at all regarding changes, I would definitely like to hear your thoughts on how it can be organized a bit better? I am sure I am missing the point there. Is the problem with the additional arguments for the BlackoilModelEbos constructor?

Reply:
Probably the issue is on my side: in my naive understanding an aquifer is just a water-only region, but based on your comments I suspect that ECL does something more crazy. is there a "two sentence executive summary" of what an ECL aquifer is?

For the issue of adding new classes, I don't mind the additional argument as I suppose the code which you propose is fine. My point is that having  a separate class for each combination of features and having N features leads to having to write 2^N classes if these features ought to be combined. even worse, these classes have to stay in sync which is IMO a direct route to maintenance hell.

Be aware that I haven't looked at the PR closely enough to determine if your "aquifer" model classes can be "bolted" on top of the regular model, in which case I would like to ask you to kindly forget my ramblings. (if so, the name of that class is a bit confusing, but again, that might be just me.) the combinatorial explosion issue is not purely theoretical rumination, though: There are already twophase, solvent and polymer (plus -- soon -- energy) extensions to flow, each of which probably ought to be combinable with aquifers.

Reply:
If there are more similar models joining, maybe we should put them in general containers based on their characteristics.